### PR TITLE
Add Result<Result<T, E>, E>::flatten -> Result<T, E>

### PR DESCRIPTION
**Pull request by Nemo157**
This PR makes this possible (modulo type inference):

```rust
assert_eq!(Ok(6), Ok(Ok(6)).flatten());
```

Tracking issue: #70142 

<sub>largely cribbed directly from <https://github.com/rust-lang/rust/pull/60256></sub>

### CODE REVIEW

File: src/libcore/result.rs [hunk 1]
```diff
@@ -1214,6 +1214,38 @@ impl<T, E> Result<Option<T>, E> {
     }
 }
 
+impl<T, E> Result<Result<T, E>, E> {
+    /// Converts from `Result<Result<T, E>, E>` to `Result<T, E>`
```
Comment (LeSeulArtichaut) [inline src/libcore/result.rs]: I think the documentation could be reworded a bit, to emphasize what happens if one of the results is an `Err` variant (which wasn't really needed for `Option<T>::flatten`). It should be noted that, in Rustdoc, you don't directly see the type of `self` when looking at the definition of a method.

**Comment by CryZe** [comment]
Should this possibly also convert the error type like the try operator does?

**Comment by Amanieu** [comment]
@Nemo157 Could you please address @CryZe's feedback:

> Should this possibly also convert the error type like the try operator does?

**Comment by Centril** [comment]
It seems strange to me that an operation named `flatten` (implying [monadic `join`](https://en.wikipedia.org/wiki/Monad_(functional_programming))) should do such implicit conversions, as `Result<T, E>` and `Result<T, F>` are not "the same monads". The equivalent `Option::flatten` doesn't, while it ostensibly could.

**Pull request merged**